### PR TITLE
[VTAdmin][vtctldserver] Keyspace Functions

### PR DESCRIPTION
**Pull request by notfelineit**
# Keyspace Actions
This PR adds the following Keyspace actions to the vtadmin UI:
- `Validate Keyspace`
- `Validate Schema`
- `Validate Version`

**Validate Keyspace**
validate_keyspace

**Validate Schema**
validateschema

**Validate Version**
validate_version

**Read Only Mode**
In read only mode, the information dropdown disappears completely:

Screen Shot 2022-02-22 at 4 38 30 PM

# Vtctldserver
This PR also adds `ValidateSchemaKeyspace` and `ValidateVersionKeyspace` to vtctldserver, and migrates the old vtctl command to use the new implementation.

**Command line clients:**
Screen Shot 2022-02-17 at 3 44 46 PM

## Testing
Two tests have been added to `server_test.go` for the `vtctldserver` library: `TestValidateSchemaKeyspace` and `TestValidateVersionKeyspace`.

```bash
> cd go/vt/vtctl/grpcvtctldserver
> go test -run TestValidateVersionKeyspace
> go test -run TestValidateSchemaKeyspace
```

To run and test the client manually:
```bash
> make build
> cd examples/local
> ./101_initial_clusters.sh
> vtctldclient ValidateVersionKeyspace commerce
> vtctldclient ValidateSchemaKeyspace commerce
```

## To Do
- [x] Support all options in `ValidateSchemaKeyspace` cobra command and main func
- [x] Port over old tests from wrangler
- [x] Point old `ValidateVersionKeyspace` and `ValidateSchemaKeyspace` to use the version defined in vtctldclient

## Related Issue(s)

- `Validate Version`: https://github.com/vitessio/vitess/issues/8733
- `Validate`: https://github.com/vitessio/vitess/issues/8735
- `ValidateSchemaKeyspace`: https://github.com/vitessio/vitess/issues/8734

## Checklist
- [x] Should this PR be backported? **No**
- [x] Tests were added or are not required **Tests were added**
- [x] Documentation was added or is not required **No new documentation needed**

## Deployment Notes

N/A

### CODE REVIEW

File: proto/vtctldata.proto
Comment (ajm188) [inline proto/vtctldata.proto:999]: missing newline here

File: proto/vtctlservice.proto
Comment (ajm188) [inline proto/vtctlservice.proto:125]: nit: punctuation (i'm _so_ sorry 🙈)

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:1392]: This should follow the form of the other method comments

File: go/vt/wrangler/schema.go
Comment (ajm188) [inline go/vt/wrangler/schema.go:111]: can we reimplement this using the new apis?

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:2891]: do we have an issue filed for this? if not, can you make one?

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:2990]: fwiw, i think a lot of these comments you copied over are superfluous and can be removed

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:3039]: can we rename these variables to make the comments unnecessary? i feel like that will make the code further down easier to read as well

**Review by ajm188 — COMMENTED** [review]
okay, I took a first pass! before we merge we should squash some of these commits down as well, I think

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (doeg) [inline web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:48]: Maybe `m-0 mb-2`?

File: web/vtadmin/src/hooks/api.ts
Comment (doeg) [inline web/vtadmin/src/hooks/api.ts:323]: Thank you for adding types to these mutations ❤️ I know that it can be kinda tedious. This sets such a great example + I appreciate it!

File: web/vtadmin/src/components/routes/Keyspaces.tsx
Comment (doeg) [inline web/vtadmin/src/components/routes/Keyspaces.tsx:98]: If it's not too much trouble, would you mind rebasing on https://github.com/vitessio/vitess/pull/9731 (when it's merged) and then hiding this column based on `isReadOnly()`? 🙇

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (doeg) [inline web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:18]: Tiny nit but FWIW I don't think we need these comments since the behaviour is well described by the function names :)

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (doeg) [inline web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:15]: Since we are both reading + writing these string values from a fixed set, I think a string enum would be helpful here. It'd give us a lil more type safety + also prevent typos.

File: go/cmd/vtctldclient/command/keyspaces.go
Comment (notfelineit) [reply go/cmd/vtctldclient/command/keyspaces.go:414]: Good call, in that case maybe the whole object should be printed (along with `ResultsByShard`)

File: web/vtadmin/src/components/toggle/Toggle.tsx
Comment (doeg) [inline web/vtadmin/src/components/toggle/Toggle.tsx:8]: So many useful components in this library! I like how you styled it. ✨

**Review by doeg — COMMENTED** [review]
The front-end changes look great! Just a couple small comments (and one medium-sized favour to ask about rebasing onto https://github.com/vitessio/vitess/pull/9731 when it's merged.) 🙇

File: go/vt/vtadmin/http/keyspaces.go
Comment (notfelineit) [reply go/vt/vtadmin/http/keyspaces.go:98]: That's what I intended (and how it works locally right now)! If query params are preferred, I can update this to use them instead.

File: proto/vtctldata.proto
Comment (notfelineit) [reply proto/vtctldata.proto:999]: makes it 1000 lines exactly ✨

File: proto/vtctlservice.proto
Comment (notfelineit) [reply proto/vtctlservice.proto:125]: No worries! Could you clarify what punctuation here?

File: go/vt/wrangler/schema.go
Comment (notfelineit) [reply go/vt/wrangler/schema.go:111]: Oh yeah! Forgot to migrate this one.

Comment (notfelineit) [reply go/vt/wrangler/schema.go:111]: Just so I understand correctly, for migration, we need to change both the vtctl command (https://github.com/vitessio/vitess/pull/9667/files#diff-ad37010accc5a29c6751d18124328bb882457e8b2527db688b854c81fe23861cR3222)  and change the replaced wrangler function?

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (notfelineit) [reply go/vt/vtctl/grpcvtctldserver/server.go:2891]: No issue - I'll file one!

Comment (notfelineit) [reply go/vt/vtctl/grpcvtctldserver/server.go:2891]: https://github.com/vitessio/vitess/issues/9732 added one here 🙂

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (notfelineit) [reply go/vt/vtctl/grpcvtctldserver/server.go:3039]: yep! 😄 I totally agree

File: web/vtadmin/package-lock.json
Comment (notfelineit) [reply web/vtadmin/package-lock.json:14379]: Absolutely! I think I was using 16, I'm not proactive enough to upgrade without being forced to 😆 
Screen Shot 2022-02-18 at 9 48 33 AM

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (notfelineit) [reply web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:46]: Screen Shot 2022-02-18 at 9 53 02 AM

it looks great!!

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (notfelineit) [reply web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:48]: Done! she's looking cute 😌

File: web/vtadmin/src/hooks/api.ts
Comment (notfelineit) [reply web/vtadmin/src/hooks/api.ts:323]: Np!! Thanks for holding the bar high!!

File: web/vtadmin/src/components/routes/Keyspaces.tsx
Comment (notfelineit) [reply web/vtadmin/src/components/routes/Keyspaces.tsx:98]: Yes 😄 Waiting for it to be merged!

File: web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx
Comment (notfelineit) [reply web/vtadmin/src/components/routes/keyspaces/KeyspaceActions.tsx:18]: 🙏 That's music to my ears haha!

**Comment by notfelineit** [comment]
@doeg I'll squash these before merging! Used to making many small commits 🙈

File: go/vt/wrangler/schema.go
Comment (ajm188) [reply go/vt/wrangler/schema.go:111]: Yeah, so `wrangler` can import an instance of `grpcvtctldserver`, and can then call the "new" method instead of having its own implementation. Similar to [this one](https://github.com/vitessio/vitess/blob/18ea5887823a33c642759b92d61b3c08233fbc35/go/vt/wrangler/keyspace.go#L57-L67)

File: go/cmd/vtctldclient/command/keyspaces.go
Comment (ajm188) [reply go/cmd/vtctldclient/command/keyspaces.go]: yeah!!

File: go/vt/wrangler/schema.go
Comment (notfelineit) [reply go/vt/wrangler/schema.go:111]: Sounds good! 😄 I'll replace both the methods! (ValidateVersion/ValidateSchema)

File: go/vt/vtadmin/http/keyspaces.go
Comment (doeg) [reply go/vt/vtadmin/http/keyspaces.go:98]: I think putting them in the request body makes sense for a PUT request :)

**Review by doeg — APPROVED** [review]
Approving for the web/vtadmin parts! I'll let @ajm188 sign off again for the golang comments. Will also merge https://github.com/vitessio/vitess/pull/9731 for rebasing as soon as CI is done 🐌

File: proto/vtctlservice.proto
Comment (ajm188) [reply proto/vtctlservice.proto:125]: you need a period. these result in godoc messages, so per [Effective Go](https://go.dev/doc/effective_go#commentary) they should be written with all the nitpicky grammar stuff 😅

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:3289]: I think we typically just stick to `m` whenever we don't need to distinguish between multiple mutexes for brevity.

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:3296]: Locking here effectively removes any benefits we are getting from concurrency, since only one goroutine can execute at a time; this means we pay the scheduler overhead without any upside. I'd suggest removing this, and lock/unlocking in each section where you need to mutate the overall `resp` data structure.

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [inline go/vt/vtctl/grpcvtctldserver/server.go:3246]: ditto on locking concurrency here. we want to lock only _after_ we make the remote call to the tablet

**Review by ajm188 — CHANGES_REQUESTED** [review]
need to fix up some things i missed on my last pass (sorry!) and then i think we're looking gooood 😎

File: doc/releasenotes/13_0_0_summary.md
Comment (notfelineit) [reply doc/releasenotes/13_0_0_summary.md:74]: Nooo I think this is a bad merge on my part -  I didn't write these.

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (notfelineit) [reply go/vt/vtctl/grpcvtctldserver/server.go:3179]: Good to know! Are you saying comments here are unnecessary in the future?

File: go/vt/wrangler/schema.go
Comment (notfelineit) [reply go/vt/wrangler/schema.go:114]: 🔪 🔪 🔪

**Review by ajm188 — APPROVED** [review]
LGTM!! 🚀

**Pull request merged**
